src/util/importJSON.ts

```diff
@@ -0,0 +1,160 @@
+
+interface ImportHtmlOptions {
+  skipRoot? : boolean,
+}
+
+/** Recursively calculate rank offset for a given position based on thoughts before and their children. Increment offset each time we pop out of context. */
+const getRankOffset = (thoughts: Block[], position?: number): number => {
+  const before = !position ? [...thoughts] : thoughts.slice(0, position)
+  const result = before.map(thought => thought.children.length > 0 ? 1 + getRankOffset(thought.children) : 1).reduce((acc, val) => acc + val, 0)
+  return position ? result : result + 1
+}
+
+/** Prepare thoughts for saving, skip root if necessary. */
+const prepare = (skipRoot: boolean | undefined, thoughts: Block[]) => {
+  if (!skipRoot) return thoughts
+  const head = _.head(thoughts)
+  if (!head) return thoughts
+  const tail = _.tail(thoughts)
+  return head.children.length > 0 ? [...head.children, ...tail] : tail
+}
+
+/** Calculate last thought of the first level, as this is where the selection will be restored to. */
+const calculateLastThoughtFirstLevel = (rankStart: number, rankIncrement: number, thoughtsJSON: Block[]) => {
+  const lastThoughtFirstLevelIndex = thoughtsJSON.length - 1
+  const lastThoughtFirstLevel = thoughtsJSON[lastThoughtFirstLevelIndex]
+  const rankOffset = lastThoughtFirstLevelIndex === 0 ? 0 : getRankOffset(thoughtsJSON, lastThoughtFirstLevelIndex)
+  return { value: lastThoughtFirstLevel.scope, rank: rankStart + rankOffset * rankIncrement }
+}
+
+/** Recursively iterate through thoughtsJSON and call insertThought for each thought individually to save it. */
+const saveThoughts = (context: Context, rank: number, rankIncrement: number, thoughtsJSON: Block[], insertThought: (value: string, context: Context, rank: number) => void) => {
+  thoughtsJSON.forEach((thought, index) => {
+    const rankOffset = index === 0 ? 0 : getRankOffset(thoughtsJSON, index)
+    insertThought(thought.scope, context, rank + rankOffset * rankIncrement)
+    if (thought.children.length > 0) {
+      saveThoughts([...context, thought.scope], rank + (rankOffset + 1) * rankIncrement, rankIncrement, thought.children, insertThought)
+    }
+  })
+}
+
+/** Return number of contexts in ThoughtJSON array. */
+const getContextsNum = (thoughts: Block[]): number => {
+  return thoughts.map(thought => thought.children.length > 0 ? 1 + getContextsNum(thought.children) : 1).reduce((acc, val) => acc + val, 0)
+}
+
+/** Calculate rankIncrement value based on rank of next sibling or its absence. */
+const getRankIncrement = (thoughtsJSON: Block[], state: State, context: Context, destThought: Child, rankStart: number) => {
+  const numContexts = getContextsNum(thoughtsJSON)
+  const destValue = destThought.value
+  const destRank = destThought.rank
+  const next = nextSibling(state, destValue, context, destRank) // paste after last child of current thought
+  const rankIncrement = next ? (next.rank - rankStart) / (numContexts || 1) : 1 // prevent divide by zero
```

Comment:
Notice that `numContexts` (`O(n)`) is always calculated even though it is only used in one branch of the ternary. How can you avoid this unnecessary calculation?

You should develop a habit of noticing where expensive functions (generally >= `O(n)`) are being used, and take steps to prevent unnecessary calls.

Reply:
Yes, you are right. I replaced `numContext` with direct `getContextsNum` call.